src/psyclone/domain/lfric/kernel/array_arg_metadata.py

```diff
@@ -0,0 +1,159 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2023-2024, Science and Technology Facilities Council
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author L. Turner, Met Office
+
+'''Module containing the ArrayArgMetadata class which captures the metadata
+associated with an array argument. Supports the creation, modification
+and Fortran output of an Array argument.
```

Comment:
Sorry to be a pain but I keep thinking we need to qualify our mentions of  'array' to make it clear it is something special (after all, a field boils down to an array and I think it could lead to confusion). Therefore, I'm wondering whether `ScalarArray` (and so `GH_SCALAR_ARRAY`) would be a better name for this concept - it signals that we really do mean something other than a field?
Something to discuss with at least @TeranIvy and possibly others.

Reply:
I think you have a point there, @arporter, as we can have arrays of fields.

Reply:
Discussed this with @mo-lottieturner and she's agreed to change it.

Reply:
does this include changing the file name from array_arg_metadata to scalar_array_arg_metadata?